packages/gatsby/src/bootstrap/load-themes/index.ts

```diff
@@ -59,12 +73,12 @@ const resolveTheme = async (
     themeDir,
     `gatsby-config`
   )
-  const theme = preferDefault(configModule)
+  const theme = preferDefault(configModule) as GatsbyConfig
 
   // if theme is a function, call it with the themeConfig
-  let themeConfig = theme
+  let themeConfig: GatsbyConfig = theme
   if (_.isFunction(theme)) {
-    themeConfig = theme(themeSpec.options || {})
+    themeConfig = theme(typeof themeSpec === `string` ? {} : themeSpec.options)
   }
```

Comment:
How about:

```suggestion
  const theme = preferDefault(configModule) as
    | GatsbyConfig
    | ((options?: IPluginRefOptions) => GatsbyConfig)

  // if theme is a function, call it with the themeConfig
  const themeConfig =
    typeof theme === `function`
      ? theme(typeof themeSpec === `string` ? {} : themeSpec.options)
      : theme
```

Reply:
Great suggestion! Github is not letting me commit it, so I'll add it manually.

Reply:
Implemented!